### PR TITLE
Stakeholder Needs extraction of User Stories

### PR DESCRIPTION
**Pull request by gregshue**
Add User Needs sdoc directory to hold User Need items so they can be referred to from Requirement items rather than copied into the USER_STORY. This is necessary since one User Need may be referenced by multiple requirement items, and one requirement item may need to reference multiple User Needs that it helps fulfill.

### CODE REVIEW

File: docs/user_needs/software_developer.sdoc
Comment (stanislaw) [inline docs/user_needs/software_developer.sdoc:20]: I am not sure if this specific user need is formulated well enough to serve as a real Developer need. What exactly it this trying to say from the software developer perspective? Is the word `threaded` referring to something like `tasked with`?

File: docs/user_needs/software_developer.sdoc
Comment (stanislaw) [inline docs/user_needs/software_developer.sdoc:2]: To justify the creation of a separate fragment for Software Developer needs, please come up with at least three compelling examples of developer needs.

Do we have at least one real software developer need in the existing documents?

(I fully support the idea of extracting the user stories as User Needs but I cannot think of enough developer needs collected in the existing documents so far.)

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:19]: This is a perfect example of a very valid User Need that fits in this fragment. Could you add two more good examples to demonstrate your proposal even stronger?

**Review by stanislaw — CHANGES_REQUESTED** [review]
I fully support the idea of documenting User Needs for the following reasons:

- INCOSE guide differentiates between Requirements and Needs. This PR is a good step towards capturing the user needs.
- The existing requirements anyway capture USER_STORIES, and this PR could simply extract all those user stories and make them actual user needs.
- I am only not sure if we have enough Developer Needs in the existing documents. If you believe we have those please share two+ examples to justify the existence of Developer Needs fragment at this point.

File: docs/user_needs/software_developer.sdoc
Comment (gregshue) [reply docs/user_needs/software_developer.sdoc:20]: Both the of the STATEMENTS in this PR were copied from USER_STORY texts in the current systems_requirements.sdoc. I don't expect any of those USER_STORY statements have been scrutinized, so I am comfortable copying them over as Draft. (BTW, I have no idea what "threaded" means here. I just replicated the existing text.)

File: docs/user_needs/software_developer.sdoc
Comment (gregshue) [reply docs/user_needs/software_developer.sdoc:2]: Personally, I don't like "Zephyr RTOS User" and "Software Developer" as terms. These needs exist outside of the consideration of Zephyr Project. I think the following highest-level categories fit better. Within each there are multiple roles and skill levels which have varying needs. I've included some examples of real existing-but-still-to-be-captured needs for each:

- Project Users (e.g., Product Manufacturers, Chip Vendors, CS/SwEng/CompEng/EE Students)
  - Includes a guide to getting started with the ecosystem;
  - Able to develop on Windows, MacOS, or Linux environments;
  - Able to submit issues to the project;
  - Licensed such that it can be freely integrated into a product without having to publish proprietary content.
- Project Contributors (e.g., Product Manufacturers, Chip Vendors, Contractors)
  - Includes a guide to contributing to the project
  - Able to submit PRs to issues;
  - Able to develop in a Docker container so that it matches the CI pipeline environment;
- Stakeholders (e.g., Regulators, Certification Test Houses, Linux Foundation, Platinum Members; Product Manufacturer Sponsors)
  - Able to restrict licensing of contributions and released content;
  - Enforce certifiable content development follows an approved SDLC;
  - Project Certifiable Release is extensible to enable reuse on proprietary HW.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:19]: See above.

**Comment by gregshue** [comment]
As a result of a private conversation with @stanislaw I will reduce the scope of this PR to focus on Zephyr RTOS User and generate a few more fundamental needs of this user. The goal is to get the (near empty) document files and initial structure into the repository.

File: docs/user_needs/software_developer.sdoc
Comment (gregshue) [reply docs/user_needs/software_developer.sdoc]: Removing Software Developer section from this PR.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:31]: I have removed the Software Developer section from this PR and added a handful of additional User Requirements from a Product Manufacturer. Justifications can be added if needed.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:46]: I like the statement, just want to mention that many embedded developers use Macs. I am 96% sure that Zephyr would not want to exclude them.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:58]: There are two different things hiding here: 

1) a more general one that we need to manage threads/tasks for parallel execution of activities.
2) preemptive scheduling can go right after the first, and the current `so that ...` part applies to this second one.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:72]: `configurable for reuse` is not understood. Let's make it more clear. Which configurability aspects you would like to bring forward?

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:84]: `Regulated devices` could be expanded on. Do you want to say:

`software that can be used in products which get certified according to the industry standards, such as IEC 61508, ISO 26262, ...` ?

It is just a suggestion. Let's simply make the statement say this more directly.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:46]: Also, in my understanding, supporting all three of Windows, Linux and Mac is a minimum requirement for claiming portability across OSes when it comes to development.

**Review by stanislaw — CHANGES_REQUESTED** [review]
This is certainly a step forward. Just some minor-to-medium comments.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:46]: I expected this would generate some discussion. The key question is whether to list each of the *three* as a separate need or combine them into one.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:58]: Agreed. We may want to break these into two distinct needs.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:72]: I'll change "reuse" to "use".

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:84]: If we are going to be more direct, then we need statements from Product Manufacturers within each regulated domains:

E.g., 
- As an Avionics Product Manufacturer, I want software that can be certified according to DO-178C;
- As a Global Consumer Wireless Product Manufacturer, I need software that can be certified as compliant with the EU RED security standard, so that I can sell product in the EU.

**Comment by gregshue** [comment]
I noticed these needs do not _depend_ on Zephyr, so I'd prefer to rename the fragment file `zephyr_rtos_user.sdoc` to be `product_manufacturer.sdoc`. We should expect a file `chip_vendor.sdoc` will come along at some point.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:58]: Hmm ... can an OS support preemptive scheduling without it being an RTOS?

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:72]: ... or delete "reuse on" ...

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (gregshue) [reply docs/user_needs/zephyr_rtos_user.sdoc:84]: As above, there may be two concepts here: Reuse vs certifiable. Certifiable presumes reuse. Do we need to list them separately when certifiable is one goal of the Zephyr Project?

File: docs/user_needs/index.sdoc
Comment (gregshue) [inline docs/user_needs/index.sdoc:15]: These needs are all independent of the Zephyr Project and exist before considering any RTOS. Rename `zephyr_rtos_user.sdoc` to files representing the specific user (e.g., `chip_vendor.sdoc`, `product_manufacturer.sdoc`, `student.sdoc`).

File: docs/user_needs/user_needs.sgra
Comment (stanislaw) [inline docs/user_needs/user_needs.sgra:23]: Change the title to `COMMENT`.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [inline docs/user_needs/zephyr_rtos_user.sdoc:35]: 1) RTOS user is ambiguous. Is it a software developer or a user of Zephyr RTOS in the context of a larger project?

2) How to quantify the word "easily"? E.g., "predictably low effort"?

3) Should the wording include "portable" or "Portability"?

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:46]: INCOSE rules advise to not include incomplete lists. Maybe the list of architectures could go to a COMMENT.

At least the three major versions should be supported: Windows, popular Linux (e.g., Ubuntu), macOS (latest release).

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:58]: For the first point, a COMMENT could maintain a ladder: `baremetal` -> `RTOS` -> `Linux`. RTOS does parallel task execution and management but is still very small and simple compared to Linux.

Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:58]: 1) "Parallel" execution is rather "Concurrent".
2) We need a glossary definition of "RTOS".
3) We do need to separate preemptive scheduling from RTOS. Extract the first part of the statement from the second. Distill the second part further if needed.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:72]: 1) Proprietary is too limited. Use Bespoke.
2) Aspects: hardware configurability (constraints that have to be passed in and settings). 
3) There is also Feature configurability. This has both enablements and settings. 

 (ISO26262 calls this "configuration", i.e., stuff that happens at configuration/build time). 

Note: Zephyr adopted two specific technologies: kconfig and device tree to enable Configurability.
4) Another aspect that needs to be captured is Reuse (Greg): Without changing a commit of Zephyr repository, we can build any product and get it certified.

File: docs/user_needs/zephyr_rtos_user.sdoc
Comment (stanislaw) [reply docs/user_needs/zephyr_rtos_user.sdoc:84]: The same as for the other comments: an incomplete list could be moved to a COMMENT and expanded on.

File: docs/user_needs/index.sdoc
Comment (stanislaw) [inline docs/user_needs/index.sdoc:1]: Consider renaming User Needs to Stakeholder needs.

**Comment by romkell** [comment]
- I fully agree to move the user story (from agile approaches) out of the existing requirement to support multiple linkage (as I stated in the req focus group meeting the user story were added to make pre-existing requirements more understandable were we had trouble to do so)
- I do not see why to formalize more than what an agile approach does
- I do not see why to split into several stakeholder. A device manufacturer, a chip vendor and a student will have 90% the same expectations of what an RTOS shall do/provide for them. There are 10% where the student does not care, where to the device manufacturer it is essential. We will get massive overlap.
- The different stakeholder needs / requests are currently gathered in github issues / RFCs in some way.
- I do not see how the user needs / requirements currently help to get together a requirements base for the existing code base that serves any qualifications (safety, security etc.).
- I am also convinced that I do not require any stakeholder needs to get a qualification (as manufacturer I can define a product/system  -with requirements - that I believe suites my stakeholders. I can develop it and certify it. Dependent on whether with my assumptions I hit the stakeholders needs, I have a certified product or a certified brick.
- If we open that door now with user needs / requirements and link it to the rest of the requirements base, we will have to manage and maintain it (also for any certification if we us it) and build a process around it. It needs to be up-to-date and consistent.
- I highly recommend to postpone to start with user needs / requirements and re-focus on creating a requirements base for the existing code base and the so far defined scope that we want to (safety) certify with the openness to be able to extend.

**Review by simhein — CHANGES_REQUESTED** [review]
I fully agree with what @romkell commented.

I would like to add that user needs are also not the goal what the the safety committee is looking for and try to achieve in this repository. Therefore I will block this PR to get moved forward and merged in my position as safety chair which stands in for the safety committees goals.

**Comment by gregshue** [comment]
Interesting comments from @romkell and @simhein. I think I can fill in some of the gaps and identify some of the incorrect assumptions. Please consider the following:

@romkell wrote
> I do not see why to formalize more than what an agile approach does

This is a wonderfully open and honest statement. Thank you! Many people confuse the Agile Manifesto statements with Scrum or XP. This approach is actually **very consistent with the Agile Manifesto once you recognize the following**:

1. Regulators have specified the definition of "working software" to include the development process and documentary evidence - and tracing out content to make sure the software does nothing beyond what is required at the system level;
2. Most customers will not be collaborating with the Zephyr Project. I am almost the only Product Manufacturer collaborating on Requirements.

One of the struggles with an approach like Scrum is that the requirements/features are not cross-linked and the process does not thoroughly examine the collection for gaps, additions, and conflicts. This has led to security holes and probably safety issues. 

> will have 90% ...

This is an assumption. Gathering (eliciting) user needs will show if that assumption is correct.

> I do not see why to split into several stakeholder.

From SEVOCAB (drawn from international standards like IEEE, ISO, IEC, etc):
> stakeholder. (1) individual or organization having a right, share, claim, or interest in a system or in its possession of characteristics that meet their needs and expectations ([ISO/IEC/IEEE 12207:2017 Systems and software engineering--Software life cycle processes](http://www.iso.org/iso/en/CatalogueListPage.CatalogueList), 3.1.59) ([ISO/IEC/IEEE 15288:2023 Systems and software engineering--System life cycle processes](http://www.iso.org/iso/en/CatalogueListPage.CatalogueList), 3.44) ([ISO/IEC/IEEE 24748-1:2018 Systems and software engineering--Life cycle management--Part 1: Guidelines for life cycle management](http://www.iso.org/iso/en/CatalogueListPage.CatalogueList), 3.51); ...

Stakeholders also include The Linux Foundation, TUV, regulators, project members, project participants, non-participating software development efforts providing integrations downstream. These span a very broad range of needs. The collection will be too large to manage as a flat list, and requirements must trace back to User Needs.

From SWEBOK v4 (public review draft):
> Formally, a software requirement has been defined as [28]:
> • a condition or capability needed by a user to solve a problem or achieve an objective;
> • a condition or capability that must be met or possessed by a system or system component to satisfy a contract, standard, specification or other formally imposed document;
> • a documented representation or capability as in (1) or (2) above.
>
> [28] ]ISO/IEC/IEEE, “ISO/IEC/IEEE 24765:2017 Systems and Software Engineering — Vocabulary,” 2nd ed. 2017

Documents get imposed on a set of software ultimately to meet User Needs.

> The different stakeholder needs / requests are currently gathered in github issues / RFCs in some way.

Agreed, which is problematic as pointed out above. That will not be sufficient to be certified as meeting the EU Cybersecurity Act or EU RED directive. I expect it will also not be sufficient to meet some safety regulations.

> I do not see how the user needs / requirements currently help

The System Requirements (the scope of the problem that will be solved) must trace up to Stakeholder Requirements (imposed constraints) and Stakeholder Needs (the breadth of the problem space that exists). IEC 62508 is introduced directly or indirectly as a Stakeholder Requirement. 

> I am also convinced that I do not require any stakeholder needs to get a qualification

You may not. Please allow that others may. AFAICT, I do.

> re-focus on creating a requirements base for the existing code base

We are actually headed to the same goal. You will see this when I fill in more content.  Here is the approximate tracing following the Requirements Engineering process. Tools will help us identify where a requirement is not met and where unintended functionality gets introduced.:

1. Stakeholder (User) Need: As a Product Manufacturer I need an RTOS to support Counting Semaphores so that I can reuse my existing RTOS-based designs.
2. System Requirement: The Zephyr Project shall produce an RTOS that provides Counting Semaphores (with tracing back to Stakeholder Requirements and/or Needs).
3. System Architecture: The Zephyr Project has a codebase containing a safe and secure RTOS providing Counting Semaphores.
...
4. Kernel Requirement: The Zephyr Kernel shall expose an interface for build-time initialization of a Counting Semaphore that ... (with tracing back to System Requirements).
5. Kernel Architecture: The Zephyr Kernel has a macro for statically instantiating and initializing a Counting Semaphore with ...
6. Kernel Architecture: The Zephyr Kernel has a component Counting Semaphore to be implemented in file sem.c that is responsible for ...
7. Interface Specification (within zephyr/kernel.h, and traced to interface requirements) ...
8. Interface Implementation (within kernel/sem.c)
9. Verification (within tests/.../kernel/sem*/ with test cases traced to interface specification.
10. Kernel File: 

> postpone to start with user needs / requirements

> with the openness to be able to extend the requirements base.

Postponing and blocking doesn't sound very open to extensions. 

From @simhein 
> user needs are also not the goal

Agreed. They are a dependency.

> safety committee is looking for and try to achieve in this repository

If the Safety Committee is trying to gather requirements independent of Security Committee needs then it isn't fulfilling the responsibilities it accepted on behalf of the whole Zephyr Project. The Zephyr Project needs one set of requirements that fulfills both needs. 

The easiest and most transparent way to see how this all fits together is by showing this with a freely available standard referenced by regulations relevant to the Zephyr Project. One example of this is ETSI 303 645, referenced by the EU Cybersecurity Act (regulation). As long as the requirements base is open to accepting content related to ETSI 303 645 then we can publicly demonstrate the entire process and solution even without submitting for security compliance approval.

**Comment by gregshue** [comment]
@romkell @simhein How about I trim this down to just the structure then migrate the User Story content into the new structure?

**Comment by gregshue** [comment]
@simhein @romkell I have trimmed this PR down to the document skeletons and extracted the User Stories from the system_requirements. Please review again.

File: docs/stakeholder_needs/index.sdoc
Comment (gregshue) [inline docs/stakeholder_needs/index.sdoc:17]: Consider placing product_manufacturer and chip_vendor within a section for Users.

File: docs/stakeholder_needs/index.sdoc
Comment (gregshue) [inline docs/stakeholder_needs/index.sdoc:22]: Consider placing project_sponsor within a section for Non-Contributing Stakeholders.

File: docs/stakeholder_needs/product_manufacturer.sdoc
Comment (gregshue) [inline docs/stakeholder_needs/product_manufacturer.sdoc:343]: This is quite long for a flat list. It needs some structure to be managable over time.

**Comment by gregshue** [comment]
Feedback from 2024 Apr 25 Req Focus Group: Reduce this to relocating the existing User Story statements in the unfragmented stakeholder_needs.sdoc, without any further organizing sections.

**Pull request closed without merging**
